<!DOCTYPE html>
<html>
<head>
  <meta charset="utf-8">
  
  
  <title>Java异常程序演示 | My Blog</title>
  <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
  <meta name="description" content="异常演示一： 12345678910111213141516171819public class test1&amp;#123;    public static void main(String[] args)&amp;#123;        A aa&#x3D;new A();        aa.divide(6,0);        System.out.println(&quot;AAAA&quot;">
<meta property="og:type" content="article">
<meta property="og:title" content="Java异常程序演示">
<meta property="og:url" content="https://username.github.io/2021/03/26/Java%E5%BC%82%E5%B8%B8%E7%A8%8B%E5%BA%8F%E6%BC%94%E7%A4%BA/index.html">
<meta property="og:site_name" content="My Blog">
<meta property="og:description" content="异常演示一： 12345678910111213141516171819public class test1&amp;#123;    public static void main(String[] args)&amp;#123;        A aa&#x3D;new A();        aa.divide(6,0);        System.out.println(&quot;AAAA&quot;">
<meta property="og:locale">
<meta property="article:published_time" content="2021-03-26T07:02:43.000Z">
<meta property="article:modified_time" content="2021-03-30T12:34:20.750Z">
<meta property="article:author" content="omnisciencepeople">
<meta name="twitter:card" content="summary">
  
    <link rel="alternate" href="/atom.xml" title="My Blog" type="application/atom+xml">
  
  
    <link rel="shortcut icon" href="/favicon.png">
  
  
    
<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/typeface-source-code-pro@0.0.71/index.min.css">

  
  
<link rel="stylesheet" href="/css/style.css">

  
    
<link rel="stylesheet" href="/fancybox/jquery.fancybox.min.css">

  
<meta name="generator" content="Hexo 5.3.0"></head>

<body>
  <div id="container">
    <div id="wrap">
      <header id="header">
  <div id="banner"></div>
  <div id="header-outer" class="outer">
    <div id="header-title" class="inner">
      <h1 id="logo-wrap">
        <a href="/" id="logo">My Blog</a>
      </h1>
      
    </div>
    <div id="header-inner" class="inner">
      <nav id="main-nav">
        <a id="main-nav-toggle" class="nav-icon"></a>
        
          <a class="main-nav-link" href="/">Home</a>
        
          <a class="main-nav-link" href="/archives">Archives</a>
        
      </nav>
      <nav id="sub-nav">
        
          <a id="nav-rss-link" class="nav-icon" href="/atom.xml" title="RSS Feed"></a>
        
        <a id="nav-search-btn" class="nav-icon" title="Suche"></a>
      </nav>
      <div id="search-form-wrap">
        <form action="//google.com/search" method="get" accept-charset="UTF-8" class="search-form"><input type="search" name="q" class="search-form-input" placeholder="Suche"><button type="submit" class="search-form-submit">&#xF002;</button><input type="hidden" name="sitesearch" value="https://username.github.io"></form>
      </div>
    </div>
  </div>
</header>

      <div class="outer">
        <section id="main"><article id="post-Java异常程序演示" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2021/03/26/Java%E5%BC%82%E5%B8%B8%E7%A8%8B%E5%BA%8F%E6%BC%94%E7%A4%BA/" class="article-date">
  <time class="dt-published" datetime="2021-03-26T07:02:43.000Z" itemprop="datePublished">2021-03-26</time>
</a>
    
  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 class="p-name article-title" itemprop="headline name">
      Java异常程序演示
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <h3>异常演示一：</h3>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br></pre></td><td class="code"><pre><span class="line">public class test1&#123;</span><br><span class="line">    public static void main(String[] args)&#123;</span><br><span class="line">        A aa&#x3D;new A();</span><br><span class="line">        aa.divide(6,0);</span><br><span class="line">        System.out.println(&quot;AAAA&quot;);</span><br><span class="line">    &#125;</span><br><span class="line">&#125;</span><br><span class="line">class A&#123;</span><br><span class="line">    int divide(int a,int n)</span><br><span class="line">    &#123;</span><br><span class="line">        return a&#x2F;n;</span><br><span class="line">    &#125;</span><br><span class="line">&#125;</span><br><span class="line">&#x2F;*</span><br><span class="line">没有输出，11行出错连带4行出错，抛出java.lang.ArithmeticException；</span><br><span class="line">11行出错divide函数无处理措施，从divide函数跳出到主函数，主函数无处理措施，从主函数跳出，将其抛给JDK中的JVM处理，JVM以终止程序的方式处理异常；</span><br><span class="line">总结：层层抛出异常直至解决异常;</span><br><span class="line">该程序不会输出AAAA;</span><br><span class="line">*&#x2F;</span><br></pre></td></tr></table></figure>
<h3>异常演示一处理</h3>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br></pre></td><td class="code"><pre><span class="line">public class test2&#123;</span><br><span class="line">    public static void main(String[] args)&#123;</span><br><span class="line">        A aa&#x3D;new A();</span><br><span class="line">        try</span><br><span class="line">        &#123;</span><br><span class="line">        aa.divide(6,0);&#x2F;&#x2F;try里面存放可能出现错误的代码</span><br><span class="line">      &#125;</span><br><span class="line">      catch(ArithmeticException e)&#x2F;&#x2F;接受异常对象</span><br><span class="line">      &#123;</span><br><span class="line">          e.printStackTrace();</span><br><span class="line">          System.out.println(&quot;除数为0&quot;);</span><br><span class="line">      &#125;</span><br><span class="line">        System.out.println(&quot;AAAA&quot;);</span><br><span class="line">    &#125;</span><br><span class="line">&#125;</span><br><span class="line">class A&#123;</span><br><span class="line">    int divide(int a,int n)</span><br><span class="line">    &#123;</span><br><span class="line">        return a&#x2F;n;</span><br><span class="line">    &#125;</span><br><span class="line">&#125;</span><br><span class="line">&#x2F;*</span><br><span class="line">若try里面的语句未出现异常，则不执行catch里面的语句；</span><br><span class="line">我们必须自己判断那些代码可能出现异常并将其放入try中;</span><br><span class="line">该程序输出：</span><br><span class="line">java.lang.ArithmeticException: &#x2F; by zero</span><br><span class="line">        at A.divide(test2.java:19)</span><br><span class="line">        at test2.main(test2.java:6)</span><br><span class="line">除数为0</span><br><span class="line">AAAA</span><br><span class="line">*&#x2F;</span><br></pre></td></tr></table></figure>
<h3>异常赋值处理</h3>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br></pre></td><td class="code"><pre><span class="line">public class test3&#123;</span><br><span class="line">    public static void main(String[] args)&#123;</span><br><span class="line">        int i;</span><br><span class="line">        try</span><br><span class="line">        &#123;</span><br><span class="line">            i&#x3D;0;</span><br><span class="line">            System.out.println(i);</span><br><span class="line">        &#125;</span><br><span class="line">        catch(Exception e)</span><br><span class="line">        &#123;</span><br><span class="line">        &#125;</span><br><span class="line">        System.out.println(i);&#x2F;&#x2F;未初始化；</span><br><span class="line">    &#125;</span><br><span class="line">&#125;</span><br><span class="line">&#x2F;*</span><br><span class="line">try 里面的赋值语句可能会因为异常而得不到执行，导致i仍未被赋值；</span><br><span class="line">*&#x2F;</span><br></pre></td></tr></table></figure>
<h3>try catch和if else关系</h3> 
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br></pre></td><td class="code"><pre><span class="line">import java.util.*;</span><br><span class="line">public class test4&#123;</span><br><span class="line">    public static void main(String[] args)&#123;</span><br><span class="line">        int i;</span><br><span class="line">        Scanner sc&#x3D;new Scanner(System.in);</span><br><span class="line">        try</span><br><span class="line">        &#123;</span><br><span class="line">            i&#x3D;sc.nextInt();</span><br><span class="line">            System.out.println(i);</span><br><span class="line">        &#125;</span><br><span class="line">        catch(Exception e)</span><br><span class="line">        &#123;</span><br><span class="line">            System.out.println(&quot;数据错误，程序终止&quot;);</span><br><span class="line">        &#125;</span><br><span class="line">    &#125;</span><br><span class="line">&#125;</span><br><span class="line">&#x2F;*</span><br><span class="line">该程序不可被if else替换来处理异常，必须用try catch</span><br><span class="line">因为我们事先不知道用户会输入什么；</span><br><span class="line">*&#x2F;</span><br></pre></td></tr></table></figure>

<h3>printStackTrace方法介绍</h3>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br></pre></td><td class="code"><pre><span class="line">class A&#123;</span><br><span class="line">    int divide(int a,int b)&#123;</span><br><span class="line">        return a&#x2F;b;</span><br><span class="line">    &#125;</span><br><span class="line">    public void f()&#123;</span><br><span class="line">        g();</span><br><span class="line">    &#125;</span><br><span class="line">    public void g()&#123;</span><br><span class="line">        divide(2,0);</span><br><span class="line">    &#125;</span><br><span class="line">&#125;</span><br><span class="line">public class test5&#123;</span><br><span class="line">    public static void main(String[] args)&#123;</span><br><span class="line">    try</span><br><span class="line">    &#123;</span><br><span class="line">      new A().f();</span><br><span class="line">    &#125;</span><br><span class="line">    catch(Exception e)</span><br><span class="line">    &#123;</span><br><span class="line">      e.printStackTrace();&#x2F;&#x2F;该方法会打印出 出错路径，便于调试</span><br><span class="line">    &#125;</span><br><span class="line">    &#125;</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>
<h3>可处理可不处理异常</h3>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br></pre></td><td class="code"><pre><span class="line">import java.io.IOException;</span><br><span class="line">class A</span><br><span class="line">&#123;</span><br><span class="line">    public void f()</span><br><span class="line">    &#123;</span><br><span class="line">        throw new IOException();&#x2F;&#x2F;抛出异常</span><br><span class="line">    &#125;</span><br><span class="line">&#125;</span><br><span class="line">public class test6&#123;</span><br><span class="line">    public static void main(String[] args)&#123;</span><br><span class="line">        new A();&#x2F;&#x2F;未调用f()方法但仍然编译报错</span><br><span class="line">    &#125;</span><br><span class="line">&#125;</span><br><span class="line">&#x2F;*</span><br><span class="line">IOException 为Exception的子类但不是RuntimeException的子类，即使不调用也必须要处理，否则编译时会出错；</span><br><span class="line">如果是异常为RuntimeException的子类，则可处理异常也可不处理异常，编译时不会出错；</span><br><span class="line">*&#x2F;</span><br></pre></td></tr></table></figure>
<h3>try···catch处理 必须处理的异常</h3>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br></pre></td><td class="code"><pre><span class="line">import java.io.IOException;</span><br><span class="line">class A</span><br><span class="line">&#123;</span><br><span class="line">    public void f()</span><br><span class="line">    &#123;</span><br><span class="line">      try</span><br><span class="line">      &#123;</span><br><span class="line">      throw new IOException();&#x2F;&#x2F;抛出异常</span><br><span class="line">      &#125;</span><br><span class="line">      catch(Exception e)</span><br><span class="line">      &#123;</span><br><span class="line">      &#125;</span><br><span class="line">    &#125;</span><br><span class="line">&#125;</span><br><span class="line">public class test7&#123;</span><br><span class="line">    public static void main(String[] args)&#123;</span><br><span class="line">        new A();</span><br><span class="line">    &#125;</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>
<h3>Throws处理 必须处理的异常</h3>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br></pre></td><td class="code"><pre><span class="line">import java.io.IOException;</span><br><span class="line">class A</span><br><span class="line">&#123;</span><br><span class="line">    public void f() throws IOException</span><br><span class="line">    &#123;</span><br><span class="line">        throw new IOException();&#x2F;&#x2F;抛出异常</span><br><span class="line">    &#125;</span><br><span class="line">&#125;</span><br><span class="line">public class test8&#123;</span><br><span class="line">    public static void main(String[] args)&#123;</span><br><span class="line">    &#125;</span><br><span class="line">&#125;</span><br><span class="line">&#x2F;*</span><br><span class="line">若main中未调用f()则编译不会报错；</span><br><span class="line">若在main调用f()则会编译会报错；</span><br><span class="line">若想调用f()而不会报错则有两种方式：</span><br><span class="line">1.在main方法中throws IOException,以此将该异常抛给JVM处理；</span><br><span class="line">2.在main中try...catch来捕获处理异常；</span><br><span class="line">*&#x2F;</span><br></pre></td></tr></table></figure>
<h3>throw Exception</h3>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br></pre></td><td class="code"><pre><span class="line">public class test9&#123;</span><br><span class="line">    public static void main(String[] args)&#123;</span><br><span class="line">        A aa&#x3D;new A();</span><br><span class="line">        try</span><br><span class="line">        &#123;</span><br><span class="line">            aa.divide(2,0);</span><br><span class="line">        &#125;</span><br><span class="line">        catch(Exception e)</span><br><span class="line">        &#123;</span><br><span class="line">            e.printStackTrace();</span><br><span class="line">        &#125;</span><br><span class="line">    &#125;</span><br><span class="line">&#125;</span><br><span class="line">class A&#123;</span><br><span class="line">    int divide(int a,int b)throws DivisorIsZeroException &#123;</span><br><span class="line">        int m;</span><br><span class="line">        if(b&#x3D;&#x3D;0)</span><br><span class="line">           throw new DivisorIsZeroException(&quot;除数不可为0&quot;);</span><br><span class="line">        else</span><br><span class="line">           m&#x3D;a&#x2F;b;</span><br><span class="line">        return m;</span><br><span class="line">    &#125;</span><br><span class="line">&#125;</span><br><span class="line">class DivisorIsZeroException extends Exception&#123;</span><br><span class="line">    public DivisorIsZeroException(String name)</span><br><span class="line">    &#123;</span><br><span class="line">        super(name);&#x2F;&#x2F;调用父类构造方法，描述该异常</span><br><span class="line">    &#125;</span><br><span class="line">&#125;</span><br><span class="line">&#x2F;*</span><br><span class="line">输出：</span><br><span class="line">DivisorIsZeroException: 除数不可为0</span><br><span class="line">        at A.divide(test9.java:18)</span><br><span class="line">        at test9.main(test9.java:6)</span><br><span class="line">*&#x2F;</span><br></pre></td></tr></table></figure>
<h3>异常总结</h3>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br></pre></td><td class="code"><pre><span class="line">class B extends Exception &#123;</span><br><span class="line">&#125;</span><br><span class="line">class A&#123;</span><br><span class="line">    public void f()throws B&#123;</span><br><span class="line">    &#125;</span><br><span class="line">&#125;</span><br><span class="line">public class test12&#123;</span><br><span class="line">    public static void main(String[] args)&#123;</span><br><span class="line">        A aa&#x3D;new A();</span><br><span class="line">        aa.f();</span><br><span class="line">        System.out.println(&quot;CCCC&quot;);</span><br><span class="line">    &#125;</span><br><span class="line">&#125;</span><br><span class="line">&#x2F;*</span><br><span class="line">B为必须处理的异常；</span><br><span class="line">该程序编译时报错，未对B处理；</span><br><span class="line">*&#x2F;</span><br></pre></td></tr></table></figure>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br></pre></td><td class="code"><pre><span class="line">class B extends Exception &#123;</span><br><span class="line">&#125;</span><br><span class="line">class A&#123;</span><br><span class="line">    public void f()throws B&#123;</span><br><span class="line">    &#125;</span><br><span class="line">&#125;</span><br><span class="line">public class test11&#123;</span><br><span class="line">    public static void main(String[] args) throws B&#123;</span><br><span class="line">        A aa&#x3D;new A();</span><br><span class="line">        aa.f();</span><br><span class="line">        System.out.println(&quot;CCCC&quot;);</span><br><span class="line">    &#125;</span><br><span class="line">&#125;</span><br><span class="line">&#x2F;*</span><br><span class="line">B为必须处理的异常；</span><br><span class="line">f()方法抛出的异常并未被触发（未throw出异常对象）</span><br><span class="line">输出CCCC</span><br><span class="line">*&#x2F;</span><br></pre></td></tr></table></figure>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br></pre></td><td class="code"><pre><span class="line">class B extends Exception &#123;</span><br><span class="line">&#125;</span><br><span class="line">class A&#123;</span><br><span class="line">    public void f()throws B&#123;</span><br><span class="line">        throw new B();</span><br><span class="line">    &#125;</span><br><span class="line">&#125;</span><br><span class="line">public class test12&#123;</span><br><span class="line">    public static void main(String[] args) throws B&#123;</span><br><span class="line">        A aa&#x3D;new A();</span><br><span class="line">        aa.f();</span><br><span class="line">        System.out.println(&quot;CCCC&quot;);</span><br><span class="line">    &#125;</span><br><span class="line">&#125;</span><br><span class="line">&#x2F;*</span><br><span class="line">B为必须处理的异常；</span><br><span class="line">f()抛出的异常被触发，程序终止；</span><br><span class="line">不会输出CCCC</span><br><span class="line">*&#x2F;</span><br></pre></td></tr></table></figure>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br></pre></td><td class="code"><pre><span class="line">class B extends Exception &#123;</span><br><span class="line">&#125;</span><br><span class="line">class A&#123;</span><br><span class="line">    public void f()throws B&#123;</span><br><span class="line">        throw new B();</span><br><span class="line">    &#125;</span><br><span class="line">&#125;</span><br><span class="line">public class test12&#123;</span><br><span class="line">    public static void main(String[] args)&#123;</span><br><span class="line">        A aa&#x3D;new A();</span><br><span class="line">        try</span><br><span class="line">        &#123;</span><br><span class="line">        aa.f();</span><br><span class="line">        &#125;</span><br><span class="line">        catch(Exception e)</span><br><span class="line">        &#123;</span><br><span class="line">        &#125;</span><br><span class="line">        System.out.println(&quot;CCCC&quot;);</span><br><span class="line">    &#125;</span><br><span class="line">&#125;</span><br><span class="line">&#x2F;*</span><br><span class="line">B为必须处理的异常；</span><br><span class="line">对异常进行捕获处理故不影响程序执行；</span><br><span class="line">输出CCCC；</span><br><span class="line">*&#x2F;</span><br></pre></td></tr></table></figure>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br></pre></td><td class="code"><pre><span class="line">class B extends RuntimeException &#123;</span><br><span class="line">&#125;</span><br><span class="line">class A&#123;</span><br><span class="line">    public void f()throws B&#123;</span><br><span class="line">    &#125;</span><br><span class="line">&#125;</span><br><span class="line">public class test12&#123;</span><br><span class="line">    public static void main(String[] args)&#123;</span><br><span class="line">        A aa&#x3D;new A();</span><br><span class="line">        aa.f();</span><br><span class="line">        System.out.println(&quot;CCCC&quot;);</span><br><span class="line">    &#125;</span><br><span class="line">&#125;</span><br><span class="line">&#x2F;*</span><br><span class="line">B为可处理可不处理异常，可不对其进行处理，编译器不会报错；</span><br><span class="line">输出CCCC；</span><br><span class="line">*&#x2F;</span><br></pre></td></tr></table></figure>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br></pre></td><td class="code"><pre><span class="line">class B extends RuntimeException &#123;</span><br><span class="line">&#125;</span><br><span class="line">class A&#123;</span><br><span class="line">    public void f()throws B&#123;</span><br><span class="line">        throw new B();</span><br><span class="line">    &#125;</span><br><span class="line">&#125;</span><br><span class="line">public class test12&#123;</span><br><span class="line">    public static void main(String[] args)&#123;</span><br><span class="line">        A aa&#x3D;new A();</span><br><span class="line">        aa.f();</span><br><span class="line">        System.out.println(&quot;CCCC&quot;);</span><br><span class="line">    &#125;</span><br><span class="line">&#125;</span><br><span class="line">&#x2F;*</span><br><span class="line">B为可处理可不处理异常；</span><br><span class="line">编译不会报错；</span><br><span class="line">运行会出错，未对抛出的B异常进行处理</span><br><span class="line">*&#x2F;</span><br></pre></td></tr></table></figure>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br></pre></td><td class="code"><pre><span class="line">class B extends RuntimeException &#123;</span><br><span class="line">&#125;</span><br><span class="line">class A&#123;</span><br><span class="line">    public void f()throws B&#123;</span><br><span class="line">        throw new B();</span><br><span class="line">    &#125;</span><br><span class="line">&#125;</span><br><span class="line">public class test12&#123;</span><br><span class="line">    public static void main(String[] args)throws B&#123;</span><br><span class="line">        A aa&#x3D;new A();</span><br><span class="line">        aa.f();</span><br><span class="line">        System.out.println(&quot;CCCC&quot;);</span><br><span class="line">    &#125;</span><br><span class="line">&#125;</span><br><span class="line">&#x2F;*</span><br><span class="line">B为可处理可不处理异常；</span><br><span class="line">编译不会报错；</span><br><span class="line">运行会出错，以终止程序的方式对抛出的B异常进行处理</span><br><span class="line">*&#x2F;</span><br></pre></td></tr></table></figure>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br></pre></td><td class="code"><pre><span class="line">class B extends RuntimeException &#123;</span><br><span class="line">&#125;</span><br><span class="line">class A&#123;</span><br><span class="line">    public void f()throws B&#123;</span><br><span class="line">        throw new B();</span><br><span class="line">    &#125;</span><br><span class="line">&#125;</span><br><span class="line">public class test12&#123;</span><br><span class="line">    public static void main(String[] args)&#123;</span><br><span class="line">        A aa&#x3D;new A();</span><br><span class="line">        try</span><br><span class="line">        &#123;</span><br><span class="line">        aa.f();</span><br><span class="line">      &#125;</span><br><span class="line">      catch(Exception e)</span><br><span class="line">      &#123;</span><br><span class="line">      &#125;</span><br><span class="line">        System.out.println(&quot;CCCC&quot;);</span><br><span class="line">    &#125;</span><br><span class="line">&#125;</span><br><span class="line">&#x2F;*</span><br><span class="line">B为可处理可不处理异常;</span><br><span class="line">以try···catch对抛出的异常进行处理，不会影响程序的执行；</span><br><span class="line">*&#x2F;</span><br></pre></td></tr></table></figure>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br></pre></td><td class="code"><pre><span class="line">class A extends Exception&#123;&#125;</span><br><span class="line">class B extends A&#123;&#125;</span><br><span class="line">class C&#123;</span><br><span class="line">    public void compare(int i,int j)throws A,B&#123;</span><br><span class="line">        if(i&gt;j)</span><br><span class="line">        throw new A();</span><br><span class="line">        else</span><br><span class="line">        throw new B();</span><br><span class="line">    &#125;</span><br><span class="line">&#125;</span><br><span class="line">public class test13&#123;</span><br><span class="line">    public static void main(String[] args)&#123;</span><br><span class="line">        C c&#x3D;new C();</span><br><span class="line">        try</span><br><span class="line">        &#123;</span><br><span class="line">            c.compare(-4,1);</span><br><span class="line">        &#125;</span><br><span class="line">        catch(B b)</span><br><span class="line">        &#123;</span><br><span class="line">            System.out.println(&quot;i&lt;j&quot;);</span><br><span class="line">        &#125;</span><br><span class="line">        catch(A a)</span><br><span class="line">        &#123;</span><br><span class="line">            System.out.println(&quot;i&gt;j&quot;);</span><br><span class="line">        &#125;</span><br><span class="line">    &#125;</span><br><span class="line">&#125;</span><br><span class="line">&#x2F;*</span><br><span class="line">捕获异常时先捕获子类异常，再捕获父类异常；</span><br><span class="line">*&#x2F;</span><br></pre></td></tr></table></figure>
      
    </div>
    <footer class="article-footer">
      <a data-url="https://username.github.io/2021/03/26/Java%E5%BC%82%E5%B8%B8%E7%A8%8B%E5%BA%8F%E6%BC%94%E7%A4%BA/" data-id="ckmpykono0000ucvp4a3p0ir5" data-title="Java异常程序演示" class="article-share-link">Teilen</a>
      
      
      
    </footer>
  </div>
  
    
<nav id="article-nav">
  
    <a href="/2021/04/06/%E6%8E%92%E5%BA%8F%E4%B9%8B%E5%BF%AB%E9%80%9F%E6%8E%92%E5%BA%8F%EF%BC%88%E7%BB%8F%E5%85%B8%E5%8D%95%E8%BD%B4%E6%8E%92%E5%BA%8F%EF%BC%89/" id="article-nav-newer" class="article-nav-link-wrap">
      <strong class="article-nav-caption">Neuer</strong>
      <div class="article-nav-title">
        
          排序之快速排序（经典单轴排序）
        
      </div>
    </a>
  
  
    <a href="/2021/03/26/Java%E5%BC%82%E5%B8%B8%E6%A6%82%E8%BF%B0/" id="article-nav-older" class="article-nav-link-wrap">
      <strong class="article-nav-caption">Älter</strong>
      <div class="article-nav-title">Java异常概述</div>
    </a>
  
</nav>

  
</article>


</section>
        
          <aside id="sidebar">
  
    

  
    

  
    
  
    
  <div class="widget-wrap">
    <h3 class="widget-title">Archiv</h3>
    <div class="widget">
      <ul class="archive-list"><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/04/">April 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/03/">March 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/01/">January 2021</a></li></ul>
    </div>
  </div>


  
    
  <div class="widget-wrap">
    <h3 class="widget-title">letzter Beitrag</h3>
    <div class="widget">
      <ul>
        
          <li>
            <a href="/2021/04/06/%E9%AA%8C%E8%AF%81%E6%8E%92%E5%BA%8F%E6%AD%A3%E7%A1%AE%E6%80%A7%E7%9A%84%E7%AE%97%E6%B3%95/">验证排序正确性的算法</a>
          </li>
        
          <li>
            <a href="/2021/04/06/%E6%8E%92%E5%BA%8F%E4%B9%8B%E5%BF%AB%E9%80%9F%E6%8E%92%E5%BA%8F%EF%BC%88%E7%BB%8F%E5%85%B8%E5%8D%95%E8%BD%B4%E6%8E%92%E5%BA%8F%EF%BC%89/">排序之快速排序（经典单轴排序）</a>
          </li>
        
          <li>
            <a href="/2021/03/26/Java%E5%BC%82%E5%B8%B8%E7%A8%8B%E5%BA%8F%E6%BC%94%E7%A4%BA/">Java异常程序演示</a>
          </li>
        
          <li>
            <a href="/2021/03/26/Java%E5%BC%82%E5%B8%B8%E6%A6%82%E8%BF%B0/">Java异常概述</a>
          </li>
        
          <li>
            <a href="/2021/03/23/java%E4%B9%8Bawt%E7%A8%8B%E5%BA%8F%E6%BC%94%E7%A4%BA/">java之awt程序演示</a>
          </li>
        
      </ul>
    </div>
  </div>

  
</aside>
        
      </div>
      <footer id="footer">
  
  <div class="outer">
    <div id="footer-info" class="inner">
      
      &copy; 2021 omnisciencepeople<br>
      Powered by <a href="https://hexo.io/" target="_blank">Hexo</a>
    </div>
  </div>
</footer>

    </div>
    <nav id="mobile-nav">
  
    <a href="/" class="mobile-nav-link">Home</a>
  
    <a href="/archives" class="mobile-nav-link">Archives</a>
  
</nav>
    


<script src="/js/jquery-3.4.1.min.js"></script>



  
<script src="/fancybox/jquery.fancybox.min.js"></script>




<script src="/js/script.js"></script>





  </div>
</body>
</html>